### PR TITLE
Fixed Linux AMQP crash and many other fixes.

### PR DESCRIPTION
**Pull request by LarryOsterman**
Several major fixes to AMQP/EventHubs

1. Return pointers to `AmqpMessage` instead of `AmqpMessage` values - this results in a noticeable performance improvement.
2. Return pointers to `EventData` instead of `EventData` values. 
3. Use atomic reference counting in uAMQP. This fixes a mysterious linux-only reference counting issue.
4. Many, many clang-tidy related fixes, including several rule-of-five fixes for AMQP classes.

Minor fixes:
1. Added cmake presets for clang and gcc on Windows. Note that these do NOT currently work due to missing compile options.
2. Moved Address Sanitizer configuration to cmakelists.txt so it works better with vcpkg cache.
3. Moved all uAMQP related types out of public and internal namespaces.

# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [ ] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [ ] Doxygen docs
- [ ] Unit tests
- [ ] No unwanted commits/changes
- [ ] Descriptive title/description
  - [ ] PR is single purpose
  - [ ] Related issue listed
- [ ] Comments in source
- [ ] No typos
- [ ] Update changelog
- [ ] Not work-in-progress
- [ ] External references or docs updated
- [ ] Self review of PR done
- [ ] Any breaking changes?

### CODE REVIEW

File: sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_value.hpp [hunk 1]
```diff
@@ -1070,14 +1061,15 @@ namespace Azure { namespace Core { namespace Amqp { namespace Models {
      * @returns The descriptor for this composite type.
      */
     AmqpValue const& GetDescriptor() const { return m_descriptor; }
-
+#if 0
```
Comment (antkmsft) [inline sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_value.hpp:1064]: Should we drop it entirely, or you still may need it in the future?

**Review by antkmsft — APPROVED** [review]
Thank you, Larry! Don't look that I am not leaving much feedback, I looked at the entire code, the change really looks good to me.
Do you think we should hide the use of shared_ptrs and make it looks like the user is dealing with the value types, i.e. `class LooksLikeValue{ private: std::shared_ptr<Impl> m_impl; }`?
Second, the places where we previously held by value, and mutability was not a problem, have you reviewed that there will be no non-expected modified value propagation over all the instances sharing the pointer, where previously such instances didn't have to worry about it, because each of them had their own copy?

File: sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_value.hpp (hunk 1 above)
Comment (LarryOsterman) [reply sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_value.hpp:1064]: I need to do a deeper cleanup pass on the change - there's a bunch of debugging artifacts there, including this one, I suspect.

**Comment by LarryOsterman** [comment]
> Thank you, Larry! Don't look that I am not leaving much feedback, I looked at the entire code, the change really looks good to me. Do you think we should hide the use of shared_ptrs and make it look like the user is dealing with the value types, i.e. `class LooksLikeValue{ private: std::shared_ptr<Impl> m_impl; }`? Second, the places where we previously held by value, and mutability was not a problem, have you reviewed that there will be no non-expected modified value propagation over all the instances sharing the pointer, where previously such instances didn't have to worry about it, because each of them had their own copy?

Actually, it's a great question and a great point. I am not a fan of the wrapper class for these specifically because of the mutable behavior. For the other AMQP types which are implemented as value classes wrapping a shared pointer, it makes sense because at the client level there's a 1:1 relationship between the client object and the shared object (the shared object may be referenced internally in AMQP which is why it's a shared object, but it's always directly tied to a client object). That is not going to be the case for AMQP/EventHubs messages.

Having said that, mutability is a real concern. So I'm changing the API signatures for the APIs which return shared pointers to return shared pointers to const elements. That way they will be immutable by default.

And since there are no clients of the API surface currently, there shouldn't be any situations where modifications matter.

**Pull request merged**